node/libs/concurrency/src/sync/prunable_mpsc/mod.rs

```diff
@@ -0,0 +1,117 @@
+//! Prunable, multi-producer, single-consumer, unbounded FIFO queue for communicating between asynchronous tasks.
+//! The pruning takes place whenever a new value is sent, based on a specified predicate.
+//! The channel also facilitates the asynchronous returning of result associated with the processing of values received from the channel.
+//!
+//! The separation of [`Sender`] and [`Receiver`] is employed primarily because [`Receiver`] requires
+//! a mutable reference to the signaling channel, unlike [`Sender`], hence making it undesirable to
+//! be used in conjunction.
+//!
+use crate::{
+    ctx, oneshot,
+    sync::{self, watch, Mutex},
+};
+use std::{collections::VecDeque, fmt, sync::Arc};
+
+#[cfg(test)]
+mod tests;
+
+/// Creates a channel and returns the [`Sender`] and [`Receiver`] handles.
+/// All values sent on [`Sender`] will become available on [`Receiver`] in the same order as it was sent,
+/// unless will be pruned before received.
+/// The Sender can be cloned to send to the same channel from multiple code locations. Only one Receiver is supported.
+///
+/// * [`T`]: The type of data that will be sent through the channel.
+/// * [`U`]: The type of the asynchronous returning of result associated with the processing of values received from the channel.
```

Comment:
T,U naming is not very informative at this point.
do you mean that T is a Request and U is a Response? If so I'd suggest Req/Resp naming.
And given that it supports bidirectional communication now (which I think is a nice idea) it is no longer a mpsc but rather just an async service, and it would benefit from changing the api altogether (and the name, of course):
* easy: instead of `send(Req) -> Receiver<Resp>`, just `call(Req) -> OrPruned<Result<Resp>>` would be better.
* not-so-easy: instead of `recv() -> (Req,Sender<Resp>)`, you can make channel() accept a Server as an argument, where Server just implements a function `handle(&mut self, Req) -> Result<Resp,Err>`

Reply:
Nice idea, but I removed the bidirectional communication, since it is not needed. Changes here: https://github.com/matter-labs/era-consensus/pull/56/commits/2209d5dc32e7c6ce4195c19101faf4290b68535c